### PR TITLE
[Feature] share api

### PR DESCRIPTION
**Pull request by maxhaensel**
# What does this PR do?

- It improves the Share Api, it work now for reflections as well
- The shareApi handles now loading-state and returns only the new User
- The shareApi handles recursive calling when a reflection has n-moments

### CODE REVIEW

File: frontend/src/api/checkUserAndShareAsset/checkUserAndShareAsset.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+      }),
+    )) as GraphQLResult<{
+      checkUserAndShareAsset: CheckShare
+    }>
+    if (res.errors) throw res.errors
+    if (res.data) return res.data.checkUserAndShareAsset
+  } catch (error) {
+    throw error.errors
+  }
+  throw new Error('somthing faild')
```
Comment (philschmid) [inline frontend/src/api/checkUserAndShareAsset/checkUserAndShareAsset.ts:22]:
```suggestion
```

File: frontend/src/api/share/share.ts [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+      if (ShareType === 'share') {
+        /*
+         when sharedUsers is empty or null, create a array and add the user
+        */
+        if (ref.sharedUsers === [] || ref.sharedUsers === null) {
+          update.sharedUsers = [user.id]
+          update.sharedUsersDetail = [{ id: user.id, email: user.email }]
+        }
+
+        if (update.sharedUsers?.includes(user.id)) {
```
Comment (philschmid) [inline frontend/src/api/share/share.ts]: Whats the point of this code?

**Review by philschmid — APPROVED** [review]
LGTM! Left two comments

File: frontend/src/api/share/share.ts [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+  Reflection: {
+    getQuery: GetReflecionsShareQuery,
+    updateMutation: updateReflection,
+  },
+}
+
+export async function shareAPI<T extends ShareItem>(
+  user: ShareUser,
+  id: string,
+  ShareType: ShareType,
```
Comment (philschmid) [inline frontend/src/api/share/share.ts]:
```suggestion
  shareType: ShareType,
```

File: frontend/src/api/checkUserAndShareAsset/checkUserAndShareAsset.ts (hunk 1 above)
Comment (maxhaensel) [reply frontend/src/api/checkUserAndShareAsset/checkUserAndShareAsset.ts:22]: Typescript want this Error :/ 

`Function lacks ending return statement and return type does not include 'undefined'.`

**Pull request merged**